### PR TITLE
Documentation of quantified type in `Expr`

### PR DESCRIPTION
**Pull request by Halbaroth**
While fixing the issue #1099, I wrote a bit of documentation for `Expr`.

### CODE REVIEW

File: src/lib/structures/expr.ml [hunk 1]
```diff
@@ -2584,7 +2610,7 @@ module Purification = struct
         mk_term (Sy.Var fresh_var) [] t.ty , add_let fresh_var t lets
 
       | _ -> (* detect ITEs *)
-        match t.xs with
+        match tm.xs with
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml]: Is this change intentional?

File: src/lib/structures/expr.ml [hunk 2]
```diff
@@ -1438,10 +1436,27 @@ and mk_forall_bis (q : quantified) =
...
         mk_forall_bis q
 
+(* Look for a substitution [sbt] of term variables such that:
+   - the domain of [sbt] is a subset of [binders];
+   - the application of [sbt] on the formula [f] produces a formula [g]
+     such that:
+     -- the set of free term variables of [g] in [binders] is strictly smaller;
+     -- the universal closures of [f] and [g] are equivalent.
+
+   For instance, let's [f] be the formula:
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml]:
```suggestion
   For instance, let [f] be the formula:
```

File: src/lib/structures/expr.ml [hunk 3]
```diff
@@ -70,11 +70,9 @@ and quantified = {
   toplevel : bool;
   user_trs : trigger list;
   binders : binders;
-  (* These fields should be (ordered) lists ! important for skolemization *)
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml:73]: I think it is helpful to document this invariant in the .ml file as well.

File: src/lib/structures/expr.ml (hunk 2 above)
Comment (bclement-ocp) [reply src/lib/structures/expr.ml]: Also [f] appears to be both the formula and a function within, is that right?

File: src/lib/structures/expr.ml [hunk 4]
```diff
@@ -1438,10 +1436,27 @@ and mk_forall_bis (q : quantified) =
         let q = {q with binders; user_trs = trs; sko_v; main = f } in
         mk_forall_bis q
 
+(* Look for a substitution [sbt] of term variables such that:
+   - the domain of [sbt] is a subset of [binders];
+   - the application of [sbt] on the formula [f] produces a formula [g]
+     such that:
+     -- the set of free term variables of [g] in [binders] is strictly smaller;
+     -- the universal closures of [f] and [g] are equivalent.
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml]: My understanding was that it we are only trying to transform implications of the form `x = e -> P(x)` into `P(e)`; is it actually more general?

File: src/lib/structures/expr.ml (hunk 1 above)
Comment (bclement-ocp) [reply src/lib/structures/expr.ml]: Given that the CI fails on this I'll guess it was not :rofl:

Comment (Halbaroth) [reply src/lib/structures/expr.ml]: Nope, there is no code change in a documentation PR ;)

File: src/lib/structures/expr.ml (hunk 4 above)
Comment (Halbaroth) [reply src/lib/structures/expr.ml]: It's a bit more general because it transforms the formula `∀ x : bool, x  -> Q(x)` into `Q(true)`.

I wrote this paragraph to explain why this transformation is valid, but I can write an exhaustive list of the transformations. I didn't think in terms of implications, which is simpler to understand indeed.

Comment (bclement-ocp) [reply src/lib/structures/expr.ml]: Right but if `x` is a boolean, `x` is actually `x = true` (: 

Having the general context where the transformation takes place is useful but I'd say explaining what the function actually does is more valuable.

File: src/lib/structures/expr.ml [hunk 5]
```diff
@@ -70,11 +70,9 @@ and quantified = {
   toplevel : bool;
   user_trs : trigger list;
   binders : binders;
-  (* These fields should be (ordered) lists ! important for skolemization *)
-  sko_v : t list;
+  sko_v : t list; (* This list has to be ordered for the skolemization. *)
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml:73]: The original comment applies to *both* `sko_v` and `sko_vty` but the new comment only mentions `sko_v`.

File: src/lib/structures/expr.ml [hunk 6]
```diff
@@ -1438,10 +1441,14 @@ and mk_forall_bis (q : quantified) =
         let q = {q with binders; user_trs = trs; sko_v; main = f } in
         mk_forall_bis q
 
+(* If [f] is a formula of the form [x = a -> P(x)] where [a] doesn't content
+   [x], this function produces the substitution { x |-> a }.
+
+   *Note*: this heuristic isn't used if the user has defined filters.
+
+   @return [None] if the formula hasn't the right form. *)
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml]: I think keeping the example as in your original rewriting was useful (or at least mention that we work with CNF so that `x = a -> P(x)` is written `x <> a \/ P(x)`).

File: src/lib/structures/expr.ml [hunk 7]
```diff
@@ -1438,10 +1441,14 @@ and mk_forall_bis (q : quantified) =
         let q = {q with binders; user_trs = trs; sko_v; main = f } in
         mk_forall_bis q
 
+(* If [f] is a formula of the form [x = a -> P(x)] where [a] doesn't content
+   [x], this function produces the substitution { x |-> a }.
+
+   *Note*: this heuristic isn't used if the user has defined filters.
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml]: Nit: the way to write bold text in odoc format is: `{b Note}`, I don't think `*Note*` works. (Doesn't matter here because this is not a doc comment, but odoc syntax is used below with `@return` so might as well be consistent)

**Review by bclement-ocp — APPROVED** [review]
